From d9d42c1553376dc5761f153860a00b1151ff9c7d Mon Sep 17 00:00:00 2001
From: udeved <artoo@cromnix.org>
Date: Mon, 18 Dec 2017 23:29:28 +0100
Subject: [PATCH] add artix branding

--- /dev/null
+++ b/src/branding/artix/branding.desc
@@ -0,0 +1,24 @@
+---
+componentName:  artix
+
+strings:
+    productName:         Artix Linux
+    shortProductName:    Artix
+    version:             rolling
+    shortVersion:        rolling
+    versionedName:       Artix Linux "rolling"
+    shortVersionedName:  Artix rolling
+    bootloaderEntryName: Artix
+
+images:
+    productLogo:         "logo.png"
+    productIcon:         "logo.png"
+    productWelcome:      "languages.png"
+
+slideshow:               "show.qml"
+
+style:
+   sidebarBackground:    "#292F34"
+   sidebarText:          "#FFFFFF"
+   sidebarTextSelect:    "#292F34"
+   sidebarTextHighlight: "#598cb8"
--- /dev/null
+++ b/src/branding/artix/show.qml
@@ -0,0 +1,54 @@
+/* === This file is part of Calamares - <http://github.com/calamares> ===
+ *
+ *   Copyright 2015, Teo Mrnjavac <teo@kde.org>
+ *
+ *   Calamares is free software: you can redistribute it and/or modify
+ *   it under the terms of the GNU General Public License as published by
+ *   the Free Software Foundation, either version 3 of the License, or
+ *   (at your option) any later version.
+ *
+ *   Calamares is distributed in the hope that it will be useful,
+ *   but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ *   GNU General Public License for more details.
+ *
+ *   You should have received a copy of the GNU General Public License
+ *   along with Calamares. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/*
+ * Slides images dimensions are 800x440px.
+ */
+
+import QtQuick 2.0;
+import calamares.slideshow 1.0;
+
+Presentation
+{
+    id: presentation
+
+    Timer {
+        interval: 20000
+        running: true
+        repeat: true
+        onTriggered: presentation.goToNextSlide()
+    }
+
+    Slide {
+        Image {
+            id: background
+            source: "squid.png"
+            width: 200; height: 200
+            fillMode: Image.PreserveAspectFit
+            anchors.centerIn: parent
+        }
+        Text {
+            anchors.horizontalCenter: background.horizontalCenter
+            anchors.top: background.bottom
+            text: ""
+            wrapMode: Text.WordWrap
+            width: root.width
+            horizontalAlignment: Text.Center
+        }
+    }
+}
